src/net/socket/udp_sock.c

```diff
@@ -28,79 +29,133 @@
-	}
+int iov_stream_make_queue(size_t len, size_t hdr_size, struct sk_buff_head *queue,
+		struct sock *sk, const struct sockaddr *sockaddr) {
+	int to_alloc = len;
 
-	skb_len = i = 0;
-	while (i < msg->msg_iovlen) {
-		data_len -= skb_len;
-		actual_len = data_len + UDP_HEADER_SIZE;
+	skb_queue_init(queue);
 
-		ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
+	while (to_alloc > 0) {
+		struct sk_buff *skb = NULL;
+		size_t actual_len = to_alloc + hdr_size;
 
-		if (ret != 0) {
+		int ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
+		if (ret < 0) {
+			skb_queue_purge(queue);
 			return ret;
 		}
+		skb_queue_push(queue, skb);
+		to_alloc -= actual_len - hdr_size;
+	}
+
+	return len;
+}
+
+int iov_dgram_make_queue(size_t len, size_t hdr_size, struct sk_buff_head *queue,
+		struct sock *sk, const struct sockaddr *sockaddr) {
 
-		assert(skb);
-		assert(skb->h.uh);
+	skb_queue_init(queue);
 
-		skb_len = 0;
+	struct sk_buff *skb = NULL;
+	size_t actual_len = len + hdr_size;
 
-		while (i < msg->msg_iovlen && skb_len + msg->msg_iov[i].iov_len <= actual_len) {
-			skb_len += msg->msg_iov[i].iov_len;
-			i++;
+	int ret = sk->o_ops->make_pack(sk, sockaddr, &actual_len, &skb);
+	if (ret < 0) {
+		return ret;
+	}
+
+	skb_queue_push(queue, skb);
+	return actual_len - hdr_size;
+}
+
+static int skb_queue_iov(struct sk_buff_head *queue, const struct iovec *iov, int iovlen, size_t header_len) {
+	struct sk_buff *skb = skb_queue_front(queue);
+	int i_iov = 0;
+	int skb_pos = 0, iov_pos = 0;
+	int ret = 0;
+
+	while (!skb_queue_end(skb, queue) && (i_iov < iovlen)) {
+		const int to_copy = min(skb->len - skb_pos, iov[i_iov].iov_len - iov_pos);
+		memcpy(skb->mac.raw + header_len + skb_pos, iov[i_iov].iov_base + iov_pos, to_copy);
+
+		iov_pos += to_copy;
+		skb_pos += to_copy;
+		ret += to_copy;
+
+		if (skb->len == skb_pos) {
+			skb = skb_queue_next(skb);
+			skb_pos = 0;
 		}
 
-		skb_buf_iovec(skb->h.uh + 1, skb_len, msg->msg_iov, msg->msg_iovlen);
+		if (iov[i_iov].iov_len == iov_pos) {
+			++i_iov;
+			iov_pos = 0;
+		}
+	}
 
-		udp_build(skb->h.uh, sock_inet_get_src_port(sk), to->sin_port,
-				skb_len + UDP_HEADER_SIZE);
+	return ret;
+}
 
-		udp4_set_check_field(skb->h.uh, skb->nh.iph);
+static int udp_get_udp_offset(struct sk_buff *skb) {
+	return (unsigned char *) (skb->h.uh) - skb->mac.raw;
+}
 
-		assert(sk->o_ops->snd_pack);
-		ret = sk->o_ops->snd_pack(skb);
+static int udp_sendmsg(struct sock *sk, struct msghdr *msg, int flags) {
+	const in_port_t sk_src = sock_inet_get_src_port(sk);
+	const struct sockaddr_in *const addr_to = (const struct sockaddr_in *)
+			(msg->msg_name ? msg->msg_name : &to_inet_sock(sk)->dst_in);
+	struct sk_buff_head queue;
+	int err;
 
-		if (0 > ret) {
-			return ret;
+	const size_t iov_data_len = iov_msg_len(msg->msg_iov, msg->msg_iovlen);
+
+	// TODO msg->msg_name could be NULL, see around. Will NULL be OK?
+	const struct sockaddr *sockaddr = (const struct sockaddr *)msg->msg_name;
```

Comment:
According to ip_make and ip_get_dest_addr, if sockaddr is NULL, address will be taken from socket.